src/goose/specs.clj

```diff
@@ -0,0 +1,117 @@
+(ns goose.specs
+  (:require
+    [goose.client :as c]
+    [goose.defaults :as d]
+    [goose.utils :as u]
+    [goose.worker :as w]
+
+    [clojure.spec.alpha :as s]
+    [clojure.spec.test.alpha :as st]
+    [clojure.string :as string]
+    [taoensso.nippy :as nippy]
+    [taoensso.carmine.connections :refer [IConnectionPool]]))
+
+; ========== Qualified Function Symbols ==============
+(defn- resolvable-fn? [func] (fn? @(resolve func)))
+(s/def ::fn-sym (s/and qualified-symbol? resolve resolvable-fn?))
+
+; ========== Redis ==============
+; Valid Redis URL patterns:
+; #1: redis://username:password@hostname:0-65353
+; #2: redis://hostname:0-65353
+(s/def :redis/url #(re-matches #"redis://.+:[0-9]{1,5}" %))
+(s/def :redis/pool-opts
+  (s/or :none #(= :none %)
+        :map #(map? %)
+        :iconn-pool #(satisfies? IConnectionPool %)))
+
+; ============== Brokers ==============
+(s/def :broker/redis
+  (s/keys :req-un [:redis/url]
+          :opt-un [:redis/pool-opts]))
+(s/def ::broker-opts
+  (s/or :redis (s/keys :req-un [:broker/redis])))
+
+; ============== Queue ==============
+(defn- unprefixed? [queue] (not (string/starts-with? queue d/queue-prefix)))
+(defn- not-protected? [queue] (not (string/includes? d/protected-queues queue)))
+(defn- len-below-1000? [queue] (< (count queue) 1000))
+(s/def ::queue
+  (s/and string? len-below-1000? unprefixed? not-protected?))
+
+; ============== Retry Opts ==============
+(s/def :retry/max-retries nat-int?)
+(s/def :retry/retry-delay-sec-fn-sym
+  (s/and ::fn-sym
+         #(pos-int? (@(resolve %) 0))))
```

Comment:
I understand the intent, but I feel like calling the retry function in a spec could lead to unexpected consequences. (What if the user's retry function performs side effects for some reason?)

Reply:
Will remove calling it & see if I can add a spec for return value to be a positive integer.

Reply:
@olttwa I think you can use [`fspec`](https://clojure.org/guides/spec#_higher_order_functions) for this.